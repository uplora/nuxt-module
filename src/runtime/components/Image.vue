<template>
  <div :class="ui.root({ class: props.ui?.root })">
    <picture :class="ui.picture({ class: props.ui?.picture })">
      ...
    </picture>
  </div>
</template>

<script lang="ts">
import type { Format, Size } from '../types/formats'
import type { ComponentSlots } from '../types/tv'
import { computed } from '#imports'
import { tv } from 'tailwind-variants'
import { useImage } from '../composables/useImage'

const uploraImage = tv({
  slots: {
    root: 'relative',
    picture: '',
  },
})

export interface UploraImageProps {
  src: string
  alt: string
  formats?: Format[]
  sizes?: Size[]
  loading?: 'lazy' | 'eager'
  ui?: ComponentSlots<typeof uploraImage>
}

export interface UploraImageEmits {
  load: [Event]
  error: [Event]
}
</script>

<script setup lang="ts">
const props = withDefaults(defineProps<UploraImageProps>(), {
  loading: 'lazy',
})

defineEmits<UploraImageEmits>()

const _image = useImage({
  id: props.src,
  formats: props.formats,
  sizes: props.sizes,
})

const ui = computed(() => uploraImage({}))
</script>
