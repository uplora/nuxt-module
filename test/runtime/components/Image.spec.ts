import type { Format, Size } from '../../../src/runtime/types/formats'
import { describe, expect, it } from 'vitest'
import Image, { type UploraImageProps } from '../../../src/runtime/components/Image.vue'
import ComponentRender from '../../component-render'

describe('image', () => {
  it.each([
    ['simple', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test' } }],
    ['with formats', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', formats: ['jpg', 'webp'] as Format[] } }],
    ['with sizes', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', sizes: [{ width: 100, descriptor: '1x' }] as Size[] } }],
    ['with lqip', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', lqip: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAMAAAA' } }],
    ['with loading', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', loading: 'eager' as const } }],
    ['with preload', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', preload: true } }],
    ['with nonce', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', nonce: '1234567890' } }],
    ['with imgAttrs', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', imgAttrs: { width: 100, height: 100 } } }],
    ['with ui', { props: { id: 'fgu1yd2ncj05iogcooz31u55', alt: 'test', ui: { root: 'w-full h-full' } } }],
  ])('render %s correctly', async (nameOrHtml: string, options: { props?: UploraImageProps }) => {
    const html = await ComponentRender(nameOrHtml, options, Image)
    expect(html).toMatchSnapshot()
  })
})
