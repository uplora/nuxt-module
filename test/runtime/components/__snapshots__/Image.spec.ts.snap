// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`image > render simple correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with formats correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <source type="image/webp"><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55/-/format/jpg" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with imgAttrs correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img width="100" height="100" src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with loading correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="eager" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with lqip correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <div class="bg-cover bg-no-repeat bg-center col-[1] row-[1]" style="background-image: url(&quot;data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAMAAAA&quot;);"></div>
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with nonce correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with preload correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with sizes correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55/-/resize/100x" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;

exports[`image > render with ui correctly 1`] = `
"<figure class="relative grid grid-cols-[100%] grid-rows-[100%] overflow-hidden w-full h-full">
  <!--v-if-->
  <picture class="aspect-3/2 col-[1] row-[1]">
    <!--v-if--><img src="https://fluxor.uplora.ru/fgu1yd2ncj05iogcooz31u55" alt="test" loading="lazy" class="w-full h-full block object-cover">
  </picture>
</figure>"
`;
